File: pisa-daemon/pkg/bpf/load.go
// Copyright 2022 SphereEx Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package bpf

import (
	"bytes"
	"encoding/binary"
	"fmt"

	// "C"
	"net"
	"syscall"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/ringbuf"
	"github.com/mlycore/log"
	"golang.org/x/sys/unix"
)

const (
	SockFilter = "sock_filter.o"
	TcPktMap   = "/sys/fs/bpf/tc/globals/my_pkt"
)

type Loader struct {
}

func (l *Loader) Load(ifaceName string, port uint16) error {
	tcMap, err := l.LoadTcPkgMap()
	// TODO: add loader to load this program to net dev
	// TODO: add port

	if err != nil {
		return err
	}

	if err := l.LoadSockFilter(ifaceName, port, tcMap); err != nil {
		return err
	}

	return nil
}

type Objs struct {
	Prog         *ebpf.Program `ebpf:"sql_filter"`
	TailProg1    *ebpf.Program `ebpf:"sql_filter_1"`
	FilterHelper *ebpf.Map     `ebpf:"filter_helper"`
	Buf          *ebpf.Map     `ebpf:"buf"`
	JmpTable     *ebpf.Map     `ebpf:"jmp_table"`
	MyPkt        *ebpf.Map     `ebpf:"my_pkt"`
	MyPktEvt     *ebpf.Map     `ebpf:"my_pkt_evt"`
}

func (l *Loader) LoadSockFilter(ifaceName string, port uint16, tcPkt *ebpf.Map) error {
	spec, err := ebpf.LoadCollectionSpec(SockFilter)
	if err != nil {
		return err
	}

	var objs Objs
	if err := spec.LoadAndAssign(&objs, nil); err != nil {
		return err
	}

	if err = objs.JmpTable.Update(uint32(0), uint32(objs.TailProg1.FD()), ebpf.UpdateAny); err != nil {
		return fmt.Errorf("jmptable err: %s", err)
	}

	if err = objs.FilterHelper.Update(uint8(0), port, ebpf.UpdateAny); err != nil {
		return fmt.Errorf("set filter port error: %v", err)
	}

	sock, err := l.OpenRawSock(ifaceName)
	if err != nil {
		return err
	}

	defer syscall.Close(sock)

	if err := syscall.SetsockoptInt(sock, syscall.SOL_SOCKET, unix.SO_ATTACH_BPF, objs.Prog.FD()); err != nil {
		return err
	}

	defer objs.Prog.Close()

	reader, err := ringbuf.NewReader(objs.MyPktEvt)
	if err != nil {
		return err
	}

	for {
		evt, query, err := l.ReadRecord(objs, reader)
		if err != nil {
			log.Warnln(err)
			continue
		}

		evt_value := EventValue{
			ClassId: l.CalcQoS(query),
		}

		if err := tcPkt.Update(&evt, &evt_value, ebpf.UpdateAny); err != nil {
			log.Warnln(err)
		}
	}
	return nil
}

func (l *Loader) OpenRawSock(ifaceName string) (int, error) {
	sock, err := syscall.Socket(syscall.AF_PACKET, syscall.SOCK_RAW|syscall.SOCK_NONBLOCK|syscall.SOCK_CLOEXEC,
		int(htons(syscall.ETH_P_ALL)))
	if err != nil {
		return -1, err
	}

	iface, err := net.InterfaceByName(ifaceName)
	if err != nil {
		return -1, err
	}

	sll := syscall.SockaddrLinklayer{
		Ifindex:  iface.Index,
		Protocol: htons(syscall.ETH_P_ALL),
	}

	if err := syscall.Bind(sock, &sll); err != nil {
		return -1, err
	}

	return sock, nil
}

func (l *Loader) LoadTcPkgMap() (*ebpf.Map, error) {
	return ebpf.LoadPinnedMap(TcPktMap, nil)
}
func (l *Loader) CalcQoS(query string) uint32 {
	//TODO: virtual database tc argument
	return 0
}

type EventKey struct {
	Seq   uint8
	Sport uint16
	Dport uint16
	Saddr uint32
	Daddr uint32
}

type EventValue struct {
	// tcp payload offset
	Offset  uint32
	PktLen  uint32
	ClassId uint32
}

// readRecord read record from ringbuf
func (l *Loader) ReadRecord(objs Objs, reader *ringbuf.Reader) (EventKey, string, error) {
	record, err := reader.Read()
	if err != nil {
		return EventKey{}, "", fmt.Errorf("reading from reader: %s", err)
	}

	var evt EventKey
	if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &evt); err != nil {
		return EventKey{}, "", fmt.Errorf("parsing ringbuf event: %s", err)
	}

	var evt_value EventValue
	if err := objs.MyPkt.Lookup(&evt, &evt_value); err != nil {
		return EventKey{}, "", fmt.Errorf("read event value: %s", err)
	}

	var (
		key     uint32
		value   uint8
		entries = objs.Buf.Iterate()
		chars   = make([]byte, evt_value.PktLen)
	)

	// PktLen contains COM_TYPE 1byte, so evt.PktLen-1 here
	for i := 0; i < int(evt_value.PktLen-1); i++ {
		entries.Next(&key, &value)
		chars = append(chars, value)
	}

	return evt, string(chars), nil
}

File: pisa-daemon/bpf/src/load.rs
// Copyright 2022 SphereEx Authors
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::convert::{TryFrom, TryInto};
use std::io::{ErrorKind, Error};
use std::net::Ipv4Addr;
use std::os::unix::io::AsRawFd;
use std::path::Path;
use aya::programs::{SocketFilter, tc, SchedClassifier, TcAttachType};
use aya::maps::{ProgramArray, HashMap, MapRefMut};
use aya::{Bpf, Pod, BpfLoader};
use socket2::{Socket, Domain, Type, Protocol};

pub enum LoadSockFilter {
    App,
    SQL,
}

impl LoadSockFilter {
    fn open_device(device: String) -> Result<Socket, Box<dyn std::error::Error>> {
        let socket = Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))?;
        socket.bind_device(Some(device.as_bytes()))?;
        Ok(socket)
    }

    fn load_app_filter<P: AsRef<Path>>(&self, device: String, path: P) -> Result<(), Box<dyn std::error::Error>> {
        let mut bpf = Bpf::load_file(path)?;
        let prog: &mut SocketFilter = bpf.program_mut("app").unwrap().try_into()?;
        prog.load()?;

        let socket = Self::open_device(device)?;
        prog.attach(socket.as_raw_fd())?;

        Ok(())
    }

}


#[derive(Debug, Clone, Copy)]
struct Endpoint {
	ip: u32,
    port: u16,
}

// Implements `Pod` to `Endpoint` as Map Key.
unsafe impl Pod for Endpoint { }

pub enum TrafficTyp {
    App,
    SQL,
}

pub const APP_ENDPOINTS_CLASSID_PIN_PATH: &str = "/sys/fs/bpf/tc/globals";
pub const APP_ENDPOINTS_CLASSID_MAP_NAME: &str = "app_endpoints_classid";

impl TrafficTyp {
    pub fn load<P: AsRef<Path>>(&self, obj_path: P, devices: &[&str]) -> Result<Bpf, Box<dyn std::error::Error>> {
        self.add_clsact(devices)?;
        match self {
            Self::App => {
                self.app(obj_path, devices)
            },

            Self::SQL => todo!()
        }
    }

    fn add_clsact(&self, devices: &[&str]) -> Result<(), Box<dyn std::error::Error>> {
        for device in devices {
            if let Err(e) = tc::qdisc_add_clsact(device) {
                if e.kind() != ErrorKind::AlreadyExists {
                    return Err(Box::new(e))
                }
            }
        }
        
        Ok(())
    }

    fn app<P: AsRef<Path>>(&self, path: P, devices: &[&str]) ->  Result<Bpf, Box<dyn std::error::Error>> {
        if !Path::new(APP_ENDPOINTS_CLASSID_PIN_PATH).exists() {
            let _ = std::fs::create_dir_all(APP_ENDPOINTS_CLASSID_PIN_PATH);
        }

        let mut bpf = BpfLoader::new().map_pin_path(APP_ENDPOINTS_CLASSID_PIN_PATH).load_file(path)?;
        let prog: &mut SchedClassifier = bpf.program_mut("classifier").unwrap().try_into()?;
        
        prog.load()?;
        for device in devices {
            prog.attach(device, TcAttachType::Egress)?;
        } 
        
        Ok(bpf)
    }

    pub fn load_app_config(&self, bpf: &mut Bpf, endpoint_classid: Vec<(config::Endpoint, u32)>) -> Result<(), Box<dyn std::error::Error>> {
        let bpf_map = bpf.map_mut(APP_ENDPOINTS_CLASSID_MAP_NAME)?;
        let mut bpf_map = HashMap::<_, Endpoint, u32>::try_from(bpf_map)?;
        for (ep, id) in endpoint_classid {
            let key = Endpoint {
                ip: ep.ip.parse::<Ipv4Addr>().unwrap().into(),
                port: ep.port,
            };

            bpf_map.insert(key, id, 0)?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use std::process::Command;

    use aya::maps::HashMap;

    use crate::load::{Endpoint, TrafficTyp};

    #[test]
    fn test_load_app_config() {
        let _ = Command::new("clang").args("-O2 -target bpf -g -c tc/app.c -o tc/app.o -I ./".split(" ")).spawn();
        let load = TrafficTyp::App;
        load.add_clsact(&vec!["lo"]).unwrap();
        let try_bpf = load.app("tc/app.o", &vec!["cni0"]);
        assert_eq!(try_bpf.is_err(), false);
        let bpf = try_bpf.unwrap();
        let mut map = HashMap::<_, Endpoint, u32>::try_from(bpf.map_mut("app_endpoints_classid").unwrap()).unwrap();
        
        let ep = Endpoint { ip: 11111, port: 8000 };
        map.insert(ep, 1_u32, 0).unwrap();

        let v = map.get(&ep, 0).unwrap();
        assert_eq!(v, 1);
    }
}
